### PR TITLE
Add support for NodeProperties

### PR DESCRIPTION
**Pull request by pjdarton**
Jenkins slave nodes can have NodeProperties set to, e.g. specify tool locations or set environment variables.
OpenStack slaves are generated dynamically from the plugin's "SlaveOptions" ... which had no support for setting NodeProperties.  The only NodeProperty set was a hard-coded environment variable NodeProperty used to pass the IP address of the node to the build.

This PR adds a new field "nodeProperties" to the slave options, allows users to specify whatever NodeProperties (that are appropriate to an OpenStack slave node) they wish, and these will be used when the plugin creates new nodes (the IP address environment variable still gets included automatically, but in addition to everything the user configured).

### CODE REVIEW

File: src/main/java/jenkins/plugins/openstack/compute/JCloudsSlave.java [hunk 1]
```diff
@@ -90,9 +93,7 @@ public JCloudsSlave(
                 labelString,
                 null,
                 new JCloudsRetentionStrategy(),
-                Collections.singletonList(new EnvironmentVariablesNodeProperty(
-                        new EnvironmentVariablesNodeProperty.Entry("OPENSTACK_PUBLIC_IP", Openstack.getAccessIpAddress(metadata))
-                ))
+                mkNodeProperties(Openstack.getAccessIpAddress(metadata), slaveOptions.getNodeProperties())
```
Comment (pjdarton) [inline src/main/java/jenkins/plugins/openstack/compute/JCloudsSlave.java]: Old code provided a hard-coded NodeProperty that set the env var OPENSTACK_PUBLIC_IP.
New code takes the configured list of NodeProperties (if any) and adds in the env var, so the node is created with both the user-specified NodeProperties plus the IP address.

File: src/test/java/jenkins/plugins/openstack/PluginTestRule.java [hunk 2]
```diff
@@ -12,21 +15,18 @@
 import hudson.model.LoadStatistics;
 import hudson.model.Node;
 import hudson.model.TaskListener;
-import hudson.model.User;
```
Comment (pjdarton) [inline src/test/java/jenkins/plugins/openstack/PluginTestRule.java]: My IDE removed this unused import when I told it to tidy things up - it can be put back if you want to keep changes to a minimum.
(Same applies to the other imports that were removed in this change)

File: src/test/java/jenkins/plugins/openstack/compute/JCloudsSlaveTest.java [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+import hudson.model.LabelFinder;
+import hudson.model.Node;
+import hudson.model.labels.LabelAtom;
+import hudson.slaves.EnvironmentVariablesNodeProperty;
+import hudson.slaves.NodeProperty;
+import jenkins.model.Jenkins;
+import jenkins.model.Nodes;
+import jenkins.model.TestJenkins;
+
+public class JCloudsSlaveTest {
```
Comment (pjdarton) [inline src/test/java/jenkins/plugins/openstack/compute/JCloudsSlaveTest.java]: There wasn't a unit-test class of this name before, and I've only implemented tests for the functionality I added, which is why the tests here _only_ check how the constructor handles various combinations of NodeProperties.

**Comment by pjdarton** [comment]
The [build failure](https://ci.jenkins.io/blue/organizations/jenkins/Plugins%2Fopenstack-cloud-plugin/detail/PR-270/1/pipeline/6) on ci.jenkins.io seems to be unrelated to my code changes.

FYI I did _many_ local builds, both of the current master and of my code changes, before I got a clean build run locally - I suspect that the tests are very flaky as it was rare to get a clean run when I built "master" locally.  I did (eventually) get a clean run of my code changes locally though, so I believe that the code changes are "good" even if it's difficult to get a clean build.
Note also: ci.jenkins.io has had a lot of problems recently - that may be a contributing factor to the build failure above.

File: src/main/java/jenkins/plugins/openstack/compute/SlaveOptions.java [hunk 4]
```diff
@@ -229,11 +248,16 @@ private Object readResolve() {
...
                 .fsRoot(_override(this.fsRoot, o.fsRoot))
                 .launcherFactory(_override(this.launcherFactory, o.launcherFactory))
+                .nodeProperties(_overrideC(this.nodeProperties, o.nodeProperties))
                 .retentionTime(_override(this.retentionTime, o.retentionTime))
                 .build()
         ;
     }
 
+    private @CheckForNull <T extends Collection<?>> T _overrideC(@CheckForNull T base, @CheckForNull T override) {
+        return override == null || override.isEmpty() ? base : override;
```
Comment (olivergondza) [inline src/main/java/jenkins/plugins/openstack/compute/SlaveOptions.java]: I suspect the need for specific override for collections is a symptom of a problematic contract of `SlaveOptions#nodeProperties`. There is a need to signal emptiness but we should have only one value for it (empty collection or `null`). I lean towards using `null` and make sure the collection is never empty.

Comment (olivergondza) [reply src/main/java/jenkins/plugins/openstack/compute/SlaveOptions.java]: Actually, scratch that. Empty collection is a valid value for overriding slave options. Ex.: Default options specify some properties and template can override by making the collection empty (effectively removing the options passed in as defaults). Empty collection should be handled as a legitimate value here.

File: src/main/java/jenkins/plugins/openstack/compute/SlaveOptions.java [hunk 5]
```diff
@@ -270,6 +295,13 @@ private Object readResolve() {
         return base;
     }
 
+    /** Returns null if our <tt>base</tt> value is the same as the <tt>def</tt>ault value. */
+    private @CheckForNull <T extends Collection<?>> T _eraseC(@CheckForNull T base, @CheckForNull T def) {
+        if (def == null || def.isEmpty()) return base;
```
Comment (olivergondza) [inline src/main/java/jenkins/plugins/openstack/compute/SlaveOptions.java]: Ditto, we do not need these override I think.

File: src/test/java/jenkins/plugins/openstack/compute/JCloudsCloudTest.java [hunk 6]
```diff
@@ -276,6 +277,7 @@ public void globalConfigMigrationFromV1() throws Exception {
         assertEquals("/tmp/jenkins", to.getFsRoot());
         assertEquals("jenkins-testing", to.getKeyPairName());
         assertThat(to.getLauncherFactory(), instanceOf(LauncherFactory.SSH.class));
+        assertThat(to.getNodeProperties(), empty());
```
Comment (olivergondza) [inline src/test/java/jenkins/plugins/openstack/compute/JCloudsCloudTest.java]: This will be `null` after the suggested change.

File: src/test/java/jenkins/plugins/openstack/JcascTest.java [hunk 7]
```diff
@@ -105,6 +105,7 @@ public void configure() {
             assertEquals("/tmp/foo", co.getFsRoot());
             assertEquals(((Integer) 42), co.getRetentionTime());
             LauncherFactory.SSH lf = (LauncherFactory.SSH) co.getLauncherFactory();
+            assertThat(co.getNodeProperties(), empty());
```
Comment (olivergondza) [inline src/test/java/jenkins/plugins/openstack/JcascTest.java]: This will be `null` after the suggested change.

Comment (olivergondza) [reply src/test/java/jenkins/plugins/openstack/JcascTest.java]: This definitely needs test coverage for declaring realistic node properties at JCasC level

File: src/test/java/jenkins/plugins/openstack/PluginTestRule.java [hunk 8]
```diff
@@ -114,10 +116,31 @@ public static SlaveOptions dummySlaveOptions() {
...
+        }
+        return b.build();
+    }
+
+    public static NodeProperty<Node> mkNodeProperty(int number) {
+        switch (number) {
+        case 1:
+            return new NodePropertyOne();
+        case 2:
+            return new NodePropertyOne();
```
Comment (olivergondza) [inline src/test/java/jenkins/plugins/openstack/PluginTestRule.java]: This should probably return different implementations.

File: src/test/java/jenkins/model/TestJenkins.java [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+     */
+    public static void setJenkinsInstance(Jenkins mockJenkinsOrNull) {
+        Jenkins.JenkinsHolder current = Jenkins.HOLDER;
+        if (!(current == null || current instanceof OurJenkinsHolder)) {
+            originalHolder = current;
+        }
+        if (mockJenkinsOrNull == null) {
+            Jenkins.HOLDER = originalHolder;
+        } else {
+            Jenkins.HOLDER = new OurJenkinsHolder(mockJenkinsOrNull);
```
Comment (olivergondza) [inline src/test/java/jenkins/model/TestJenkins.java]: Ouch, do we ABSOLUTELY need this? I failed to follow which part is problematic to stub without this.

File: src/test/java/jenkins/plugins/openstack/PluginTestRule.java (hunk 8 above)
Comment (pjdarton) [reply src/test/java/jenkins/plugins/openstack/PluginTestRule.java]: Oops.  Yup - good spot...

File: src/main/java/jenkins/plugins/openstack/compute/SlaveOptions.java (hunk 4 above)
Comment (pjdarton) [reply src/main/java/jenkins/plugins/openstack/compute/SlaveOptions.java]: I mostly agree ... but I'm not sure how, via the configuration WebUI, the user can distinguish between an empty list and a null list.  Unless we can let the user distinguish between the two cases, I think that the plugin functionality should treat them the same.

The jelly code is in [SlaveOptions/config.jelly line 68](https://github.com/jenkinsci/openstack-cloud-plugin/pull/270/files#diff-b4cd7ab6e132cf606fb8931efff7fd35R68) and, as far as I am aware, this gives the user no control over whether it results in an empty list or a null list.

If you can help (or get someone else to help...) work out the correct jelly magic that'll give the user control over null/empty then I'd be happy to remove the "treat empty the same as null" code.

Comment (olivergondza) [reply src/main/java/jenkins/plugins/openstack/compute/SlaveOptions.java]: Good point. I am afraid we have to enclose that with `optionalBlock` that, when unchecked would mean _I do not contribute/override any options_, and when checked it would be an explicit list of options to provide/override. That would enable the semantics of both `null` (checkbox unchecked) and empty (checked but no implementations added).

---

I do not quite like the complexity but I see no other way to permits overriding templates to erase inherited defaults.

File: src/test/java/jenkins/model/TestJenkins.java (hunk 9 above)
Comment (pjdarton) [reply src/test/java/jenkins/model/TestJenkins.java]: This is probably due to ignorance on my part, but the reason was that the `JCloudsSlave` constructor calls `AbstractCloudSlave` calls `Slave` which calls:
1. `Jenkins.getInstance().getNode(...)`
1. `Jenkins.getInstance().getLabelAtom(...)`
1. `Jenkins.getInstanceOrNull().getExtensionList(LabelFinder.class)`

I don't know how to get a "test Jenkins" set up (this aspect of Jenkins plugins has always been a mystery to me) so I mocked it.
If you know of a way it can be set up (it doesn't need to do anything other than "not throw an exception") then we could remove that test code.

TL;DR: I need the `Slave` constructor to not crash when called, which means `Jenkins.getInstance()` needs to be stubbed/mocked somehow; I'm open to suggestions on how best to do that.

File: src/main/java/jenkins/plugins/openstack/compute/SlaveOptions.java (hunk 4 above)
Comment (pjdarton) [reply src/main/java/jenkins/plugins/openstack/compute/SlaveOptions.java]: Can you point me at an example of where this is done in existing code?
I'm familiar with the _java_ `Optional` but doing this with Jenkins jelly code is unknown territory for me...

File: src/test/java/jenkins/plugins/openstack/JcascTest.java (hunk 7 above)
Comment (pjdarton) [reply src/test/java/jenkins/plugins/openstack/JcascTest.java]: OK, understood ... but I'll hold off making any additions here until we've sorted out the null/empty node properties issue on SlaveOptions.

File: src/main/java/jenkins/plugins/openstack/compute/SlaveOptions.java (hunk 4 above)
Comment (pjdarton) [reply src/main/java/jenkins/plugins/openstack/compute/SlaveOptions.java]: I've been googling...  from what I can tell, the jelly `optionalBlock` is controlled by a boolean, which I guess we'd call `hasNodeProperties`.
So, I figure there's three possibilities:
1. (worst case) We need to _store_ that boolean, i.e. we add a new boolean field `hasNodeProperties`, together with all the getters/setters that entails, and then ensure that getNodeProperties returns null if hasNodeProperties is false.
1. As above, but we can get away with it being a transient boolean whose value is initialised as nodeProperties!=null.
1. (best case) We just implement getHasNodeProperties and setHasNodeProperties to give the jelly something to talk to ... and rely on setHasNodeProperties always being called before setNodeProperties.

WDYT?

File: src/test/java/jenkins/model/TestJenkins.java (hunk 9 above)
Comment (olivergondza) [reply src/test/java/jenkins/model/TestJenkins.java]: If we only need this not to blow up, running the test with `@Rule JenkinsRule` would do. most tests in the plugin does so. It runs with real Jenkins running in that JVM

Comment (pjdarton) [reply src/test/java/jenkins/model/TestJenkins.java]: OK, I'll give that a try.

File: src/main/java/jenkins/plugins/openstack/compute/SlaveOptions.java (hunk 4 above)
Comment (olivergondza) [reply src/main/java/jenkins/plugins/openstack/compute/SlaveOptions.java]: I would have to play with that to see whether that is implementable without too much of a hassle, but we can pass the optional block an explicit condition of `checked="${it.nodeProperties != null}"` so there should be no need to maintain the boolean property alongside of the properties list. (The submitted input value will be just ignored by the `@DataBoundConstructor`).

Let me know if you prefer me to prototype that.

Comment (pjdarton) [reply src/main/java/jenkins/plugins/openstack/compute/SlaveOptions.java]: All help would be gratefully accepted :grinning: 
FYI I only implemented this because I wanted https://github.com/jenkinsci/wsclean-plugin to have a "don't clean this node" nodeProperty - what was meant to be a "fairly quick" enhancement to wsclean-plugin has snowballed out of control as I had to also add NodeProperty support to the vSphere plugin as well as this plugin (vSphere was easy compared to this one though).

File: src/test/java/jenkins/model/TestJenkins.java (hunk 9 above)
Comment (pjdarton) [reply src/test/java/jenkins/model/TestJenkins.java]: Today I learned something :grin: 
Nasty hacky test class removed in https://github.com/jenkinsci/openstack-cloud-plugin/pull/270/commits/e0ebacc677cb89350fe1ee461973ef10cb3a1495

File: src/test/java/jenkins/plugins/openstack/PluginTestRule.java (hunk 8 above)
Comment (pjdarton) [reply src/test/java/jenkins/plugins/openstack/PluginTestRule.java]: Fixed in https://github.com/jenkinsci/openstack-cloud-plugin/pull/270/commits/e0ebacc677cb89350fe1ee461973ef10cb3a1495

File: src/test/java/jenkins/plugins/openstack/JcascTest.java (hunk 7 above)
Comment (pjdarton) [reply src/test/java/jenkins/plugins/openstack/JcascTest.java]: Initial tests added as part of https://github.com/jenkinsci/openstack-cloud-plugin/pull/270/commits/f69471d6380ee1ddea2cefcc61b000c95f03484e
Note: Those changes don't test our ability to specify an empty list of node properties to override the defaults, and I'm not 100% sure of what the jcasc syntax would be for that...
My guess is that it'd be
```
          - name: "noNodeProperties"
            labels: "templateWithNoNodePropertiesInsteadOfDefaults"
            slaveOptions:
              nodeProperties: []
```
Would you agree?

File: src/main/java/jenkins/plugins/openstack/compute/SlaveOptionsDescriptor.java [hunk 10]
```diff
@@ -483,4 +484,18 @@ public FormValidation doCheckFsRoot(
...
+     * {@link JCloudsSlave}s that are created from these options.
+     *
+     * @return the filtered list
+     */
+    @Nonnull
+    @Restricted(NoExternalUse.class) // used by Jelly EL only
+    public List<NodePropertyDescriptor> getNodePropertiesDescriptors() {
+        final Jenkins j = Jenkins.get();
+        final JCloudsSlaveDescriptor jcsd = (JCloudsSlaveDescriptor) j.getDescriptor(JCloudsSlave.class);
+        return jcsd.nodePropertyDescriptors(null);
```
Comment (pjdarton) [inline src/main/java/jenkins/plugins/openstack/compute/SlaveOptionsDescriptor.java]: When I do a local build, I see that findbugs is showing
```
[INFO] Possible null pointer dereference in jenkins.plugins.openstack.compute.SlaveOptionsDescriptor.getNodePropertiesDescriptors() due to return value of called method [jenkins.plugins.openstack.compute.SlaveOptionsDescriptor, jenkins.plugins.openstack.compute.SlaveOptionsDescriptor] Dereferenced at SlaveOptionsDescriptor.java:[line 499]Known null at SlaveOptionsDescriptor.java:[line 498] NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE
```

I've checked the code and it's correct - `getDescriptor(...)` is annotated with `@CheckForNull`  - _but_ it should never return null in this case.
If it does return null when given `JCloudsSlave.class` as an argument, our plugin (and probably Jenkins as a whole) is likely to be broken for other reasons, so it probably doesn't matter if this method throws a NPE at that point :wink: 
I've googled and found that I could add code saying
```
    @edu.umd.cs.findbugs.annotations.SuppressFBWarnings(
            value="NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE", 
            justification="It's ok to NPE if *our* slave descriptor is missing")
```
in front of the method and that'll then prevent this findbugs complaint from appearing.
WDYT?

File: src/test/java/jenkins/plugins/openstack/JcascTest.java (hunk 7 above)
Comment (olivergondza) [reply src/test/java/jenkins/plugins/openstack/JcascTest.java]: Yes, that is correct. The think is that everyone's favorite markup format, YAML, has no means of representing empty lists. So the fact YAML is a superset of JSON (yes, you can switch to JSON notation at will), comes to rescue here as JSON do have the empty list literal.

File: src/main/java/jenkins/plugins/openstack/compute/SlaveOptionsDescriptor.java (hunk 10 above)
Comment (olivergondza) [reply src/main/java/jenkins/plugins/openstack/compute/SlaveOptionsDescriptor.java]: Better use `#getDescriptorOrDie()` that would throw an exception.

Comment (pjdarton) [reply src/main/java/jenkins/plugins/openstack/compute/SlaveOptionsDescriptor.java]: Fixed in https://github.com/jenkinsci/openstack-cloud-plugin/pull/270/commits/73df874c9fbedb780b913328ca0f577dfeeb73e4

**Comment by olivergondza** [comment]
@pjdarton, I managed to implement the nullability of the properties. Please pull that commit[1] in.

[1] https://github.com/olivergondza/openstack-cloud-plugin/commit/e2ab58adbed4516373e18afe01646033f12e0709

**Comment by pjdarton** [comment]
OK.  Will do (and will include JCasc test as well).

Thanks for figuring out the Jelly stuff - I don't think I would have easily figured out the 
[SlaveOptionsDescriptor.newInstance](https://github.com/olivergondza/openstack-cloud-plugin/commit/e2ab58adbed4516373e18afe01646033f12e0709#diff-f096d2662af43a89ec56d82914564ef4R507) trick on my own.

File: src/test/java/jenkins/plugins/openstack/JcascTest.java (hunk 7 above)
Comment (pjdarton) [reply src/test/java/jenkins/plugins/openstack/JcascTest.java]: Tests included in https://github.com/jenkinsci/openstack-cloud-plugin/pull/270/commits/5456d1d835b8da5588b9f7e3ce96730cc0b1a071

File: src/test/java/jenkins/plugins/openstack/compute/JCloudsCloudTest.java (hunk 6 above)
Comment (pjdarton) [reply src/test/java/jenkins/plugins/openstack/compute/JCloudsCloudTest.java]: Done in https://github.com/jenkinsci/openstack-cloud-plugin/pull/270/commits/5456d1d835b8da5588b9f7e3ce96730cc0b1a071

File: src/main/java/jenkins/plugins/openstack/compute/SlaveOptions.java (hunk 5 above)
Comment (pjdarton) [reply src/main/java/jenkins/plugins/openstack/compute/SlaveOptions.java]: Done in https://github.com/jenkinsci/openstack-cloud-plugin/pull/270/commits/5456d1d835b8da5588b9f7e3ce96730cc0b1a071

File: src/main/java/jenkins/plugins/openstack/compute/SlaveOptions.java (hunk 4 above)
Comment (pjdarton) [reply src/main/java/jenkins/plugins/openstack/compute/SlaveOptions.java]: I've included your changes (olivergondza@e2ab58a) in https://github.com/jenkinsci/openstack-cloud-plugin/pull/270/commits/5456d1d835b8da5588b9f7e3ce96730cc0b1a071
I've done some basic testing locally and it looks like it works exactly as expected :grin:

**Comment by pjdarton** [comment]
OK, as of https://github.com/jenkinsci/openstack-cloud-plugin/pull/270/commits/5456d1d835b8da5588b9f7e3ce96730cc0b1a071 I think I've addressed all the changes we've discussed.

- Setting node properties in the cloud settings causes them to be used in all templates that don't have node properties.
- Settings some node properties in a template, even if they're an empty list of properties, overrides the cloud defaults.

I think we're good to go now ... unless there's anything else we've missed?

File: src/main/java/jenkins/plugins/openstack/compute/JCloudsSlave.java (hunk 1 above)
Comment (olivergondza) [reply src/main/java/jenkins/plugins/openstack/compute/JCloudsSlave.java]: Re-reviewing this, are we making sure individual slaves will get detached instance of individual properties? IOW, the same property object does not get attached to more than one slave?

Comment (pjdarton) [reply src/main/java/jenkins/plugins/openstack/compute/JCloudsSlave.java]: No, we are not sure of that.
The list of properties is detached.
The individual properties (aside from the environment variables property, which gets special treatment) are copied by reference.

FYI the docker plugin uses copy by reference too ... and that seems to work ok like that.
One would hope that node properties are immutable once created so there should be no problems...

...also, I don't know of a generic means of making a fully independent copy of a node property.
I guess it might be possible to serialise to XML and then de-serialise, but I do not think we can assume they'll all support clone or have a copy constructor.

If it's possible to write a method that takes a NodeProperty and returns a deep copy it then that _would_ be useful ... and that'd be useful for the docker plugin and the vSphere plugin too ... but _I_ don't know how that could be done :-/

Comment (pjdarton) [reply src/main/java/jenkins/plugins/openstack/compute/JCloudsSlave.java]: I've been experimenting...
I've come up with the following code:
```
    @SuppressWarnings("unchecked")
    private static <T> T makeCopy(final T templateNP) {
        final String templateNPAsXml = Jenkins.XSTREAM.toXML(templateNP);
        final Object fromXml = (T) Jenkins.XSTREAM.fromXML(templateNPAsXml);
        final T copyOfTemplateNP = (T) fromXml;
        return copyOfTemplateNP;
    }
```

I could put that code into JCloudsSlave.java and then, where the `mergeNodeProperties` method says
```
            } else {
                b.add(templateNP);
            }
```
we'd change it to be
```
            } else {
                final NodeProperty<? extends Node> copyOfTemplateNP = makeCopy(templateNP);
                b.add(copyOfTemplateNP);
            }
```

...and then fix the corresponding unit tests (which are looking for `sameInstance`) and add another test that ensures that the node properties are still `equalTo(...)` but `not(sameInstance(...))`.

@olivergondza WDYT?

Comment (pjdarton) [reply src/main/java/jenkins/plugins/openstack/compute/JCloudsSlave.java]: ...or, in order to avoid making a new String each time, we could do:
```
    @SuppressWarnings("unchecked")
    private static <T> T makeCopy(final T original) {
        class IOStream extends ByteArrayOutputStream {
            public ByteArrayInputStream getCurrentContent() {
                return new ByteArrayInputStream(buf, 0, count);
            }
        }
        final IOStream xml = new IOStream();
        Jenkins.XSTREAM.toXML(original, xml);
        final Object copy = Jenkins.XSTREAM.fromXML(xml.getCurrentContent());
        return (T) copy;
    }
```
I can't tell for sure which would be more efficient in reality - in theory, this new version avoids the needless creation of a big String, which _should_ be quicker...

Alternatively, it might be better/quicker to use the `Reader`/`Writer` equivalents (by having the class extend `CharArrayWriter` and return a `CharArrayReader`).

However the original String method makes for much simpler code, even if it might not be the fastest.

Personally, I think that "easily read code" wins over "theoretically faster"...

Comment (olivergondza) [reply src/main/java/jenkins/plugins/openstack/compute/JCloudsSlave.java]: > Personally, I think that "easily read code" wins over "theoretically faster"...

100% agree.  Let's start with that.

Comment (pjdarton) [reply src/main/java/jenkins/plugins/openstack/compute/JCloudsSlave.java]: Implemented in https://github.com/jenkinsci/openstack-cloud-plugin/pull/270/commits/0b13f412d6251ee6cd92f498cf509f4e986e3520 ... along with a new unit-test to ensure "equalTo and not same instance"

**Comment by olivergondza** [comment]
I am releasing this as 2.48
